packages/api/src/command/medical/document/process-doc-query-webhook.ts

```diff
@@ -0,0 +1,117 @@
+import { toDTO, DocumentReferenceDTO } from "../../../routes/medical/dtos/documentDTO";
+import { getAllWebhookRequestByRequestId } from "../../webhook/webhook-request";
+import { Config } from "../../../shared/config";
+
+const { log } = out(`Doc Query Webhook`);
+const isSandbox = Config.isSandbox();
+
+/**
+ * Processes the document query progress to determine if when to send the document download and conversion webhooks
+ */
+export const processDocQueryProgressWebhook = async ({
+  patient,
+  documentQueryProgress,
+  requestId,
+}: {
+  patient: Pick<Patient, "id" | "cxId" | "externalId">;
+  documentQueryProgress: DocumentQueryProgress;
+  requestId: string;
+}): Promise<void> => {
+  const { id: patientId } = patient;
+
+  try {
+    const webhooks = await getAllWebhookRequestByRequestId(requestId);
+
+    await handleDownloadWebhook(webhooks, patient, requestId, documentQueryProgress);
+    await handleConversionWebhook(webhooks, patient, requestId, documentQueryProgress);
+  } catch (error) {
+    const msg = `Error on processDocQueryProgressWebhook`;
+    const extra = {
+      documentQueryProgress,
+      requestId,
+      patientId,
+      msg,
+      context: `webhook.processDocQueryProgressWebhook`,
+      error,
+    };
+
+    log(`${msg}: ${errorToString(error)} - ${JSON.stringify(extra)}`);
+    capture.error(error, { extra });
+  }
+};
+
+const handleDownloadWebhook = async (
+  webhooks: WebhookRequest[],
+  patient: Pick<Patient, "id" | "cxId" | "externalId">,
+  requestId: string,
+  documentQueryProgress: DocumentQueryProgress
+): Promise<void> => {
+  const downloadWebhookType = "medical.document-download";
+  const downloadStatus = documentQueryProgress.download?.status;
+  const isDownloadFinished = downloadStatus === "completed" || downloadStatus === "failed";
+  const downloadWebhookSent = webhooks.some(webhook => webhook.type === downloadWebhookType);
+
+  const canProcessRequest = isDownloadFinished && !downloadWebhookSent;
+
+  if (canProcessRequest && !isSandbox) {
+    const downloadIsCompleted = downloadStatus === "completed";
+    const payload = await composeDocRefPayload(patient.id, patient.cxId, requestId);
+
+    processPatientDocumentRequest(
+      patient.cxId,
+      patient.id,
+      downloadWebhookType,
+      downloadIsCompleted ? MAPIWebhookStatus.completed : MAPIWebhookStatus.failed,
+      requestId,
+      downloadIsCompleted ? payload : undefined
+    );
+  }
+};
+
+const handleConversionWebhook = async (
+  webhooks: WebhookRequest[],
+  patient: Pick<Patient, "id" | "cxId" | "externalId">,
+  requestId: string,
+  documentQueryProgress: DocumentQueryProgress
+): Promise<void> => {
+  const convertWebhookType = "medical.document-conversion";
+  const convertStatus = documentQueryProgress.convert?.status;
+  const isConvertFinished = convertStatus === "completed" || convertStatus === "failed";
+  const convertWebhookSent = webhooks.some(webhook => webhook.type === convertWebhookType);
+
```

Comment:
instead of doing two calls to webhooks.some of O(n), do one call and then something like this:

```
  const webhookSet = new Set(webhooks.map(webhook => webhook.type));
  const downloadWebhookSent = webhookSet.has(downloadWebhookType);
  const convertWebhookSent = webhookSet.has(convertWebhookType);
```

Reply:
Wouldnt be the same if you are then using `.has` twice?